connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/Checkpoint.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.mirror;
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.protocol.types.Field;
+import org.apache.kafka.common.protocol.types.Schema;
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.protocol.types.Type;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+
+import java.util.Map;
+import java.util.HashMap;
+import java.nio.ByteBuffer;
+
+public class Checkpoint {
```

Comment:
Have you considered using the protocol generator framework that is available for clients or would it make this more complicated than necessary?
As I see you're simply using these messages as payload and they're not really protocol messages but we might gain something as they generate hashcode and equals methods.
Also, why are these (Checkpoint, heartbeat) not protocol messages?

Reply:
As you say, these aren't really protocol messages, as they are not requests or responses, so I don't think the generator stuff would be a good fit here. Maybe there are parts of it I could use.

I suppose we could use JSON here as well. These are simple, unstructured records, so there isn't much to be gained from encoding in JSON, but it would be nice to get rid of some of this code.